src/renderers/dom/client/MarkupMismatchError.js

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule MarkupMismatchError
+ */
+
+'use strict';
+
+var getInstanceDisplayName = require('getInstanceDisplayName');
+var NativeNodes = require('NativeNodes');
+
+// we cannot use class MarkupMismatchError extends Error {} because Babel cannot extend built-in
+// objects in a way that allows for instanceof checks.
+// this error subclassing code is copied and modified from
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
+// According to https://developer.mozilla.org/en-US/docs/MDN/About#Copyrights_and_licenses, all code in MDN
+// added on or after August 20, 2010 is public domain, and according to the wiki page history, the earliest
+// version of this code was added on June 7, 2011:
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error$revision/62425
+function MarkupMismatchError(message, node, serverVersion, clientVersion) {
+  this.name = 'MarkupMismatchError';
+  this.message = message || 'The pre-rendered markup did not match the component being rendered.';
+  this.node = node;
+  this.serverVersion = serverVersion;
+  this.clientVersion = clientVersion;
+  this.stack = (new Error()).stack;
+}
+MarkupMismatchError.prototype = Object.create(Error.prototype);
```

Comment:
We explicitly stopped using `Object.create` to simplify our sham story. We don't rely on shams and it is not available in all browsers we support.

We also don't throw errors with custom classes anywhere else so if were to start, that would be a bigger discussion.

Currently our pattern is `invariant(false, 'Error message with meta data 1: %s and meta data 2: %s', metaData1, metaData2);` which allows it to be passed to a logger as rich data or just as a string but not relied upon in production code which allow us to strip it out into a very compact error.

I'd rather see us use the existing pattern and if that's not enough follow up in #2686 and we'll change it all at once.